src/messaging/ExchangeContext.cpp

```diff
@@ -412,6 +425,32 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     }
 
 exit:
+    // Don't close ourselves if we're already closed.
+    //
+    // Don't close ourselves if this message is a standalone ack. We're still
+    // not closed and getting an ack should not affect that.  In particular,
+    // since the standalone ack was not passed to the delegate, the delegate
+    // never got a chance to say "stay open". The one exception here is if
+    // mDelegate is null: in that case this is an unsolicited message and we
+    // were just created to ack it and close after that.
+    //
+    // Don't close for duplicates for similar reasons, with the same exception.
+    //
+    // Also don't close if there's an outer HandleMessage invocation.  It'll
+    // deal with the closing.
+    if (!mFlags.Has(Flags::kFlagClosed) && !mFlags.Has(Flags::kFlagWillSendMessage) && !IsResponseExpected() &&
+        (!isStandaloneAck || (mDelegate == nullptr)) && (!isDuplicate || (mDelegate == nullptr)) && !alreadyHandlingMessage)
```

Comment:
Could you explain more bout these two conditions? (!isStandaloneAck || (mDelegate == nullptr)) && (!isDuplicate || (mDelegate == nullptr))

We hit them only when OnMessageReceived return error.

Reply:
Hmm.  I tried to explain in the comments in the code; if those are not clear enough I need to work on that and would appreciate wording suggestions.

Those cases are meant to detect the case when we create an exchange just to ack a message that does not have a UMH.  In that situation we will create an exchange with a null delegate and call HandleMessage on it.  In that situation, we want to close the exchange even if the message is an ack or a duplicate.

We could simplify things a bit perhaps by factoring out the "send an ack" codepath so that it can be invoked directly in the "just send an ack" case instead of conflating it with the generic HandleMessage codepath.  That would allow simpler logic here, at the cost of resurrecting the "send ack and close" logic in the exchange manager.

Now maybe we could have a situation where someone creates an exchange with a null delegate, sends a message on it, doesn't call `WillSendMessage` to indicate the intent to send more messages and doesn't say a response is expected (and such a response would not be seen anyway, because of the null delegate).  But in such a situation, I'd think closing the exchange on receipt of the ack _is_ in fact the right course of action: semantically there is no plan to send more messages on it.

Reply:
I understand we don't close the exchange on receipt of the standalong ack, since we expect the exchange will be closed on receipt of the response message later.

Why we don't close the exchange on duplicate message? who are going to close the exchange after we switch to auto-close pattern.

Reply:
The idea is to close the exchange on application-level events like "last message of conversation sent" or "last message of conversation received".  The exchange knows when messages are sent or received; the app needs to indicate whether it's the "last" one.  For receipt it indicates this by flagging whether it plans to send a response, or just sending a response.  For send it does this via the "response expected" flag.

Duplicate message receipt is not an application-level event.  The only reason duplicates are even delivered to an exchange is so that we can generate a MRP ack for them.  

One thing we could consider doing to simplify the logic here is to treat duplicates like we treat unsolicited messages with no handler and just create a separate exchange to ack them.  But I'd rather do that sort of thing in a separate PR, especially because I think that setup is actually broken in some cases, so would need some careful analysis.

Reply:
That make sense to me, only the application who understand the protocol context has the knowledge of the last message of the conversation. This concept does not change with this PR,  the only difference is the application needs to take actions to explicitly indicate to extend the exchange, otherwise the exchange will be closed automatically by itself.

I understand we don't need to deliver duplicate message to application and only do MRP ack within exchange, my question is who finally close the exchange after MRP ack for duplicate message?

Reply:
> my question is who finally close the exchange after MRP ack for duplicate message?

Ah, I see.  There are two cases here:

1. The exchange is being used for an actual application-level conversation.  In that case, the exchange will be closed in the normal course of that conversation.
2. The exchange was created just to ack the duplicate.  In that case, it will have a null delegate and the auto-closing will happen; hence the form of the `(!isDuplicate || (mDelegate == nullptr))` condition.

Reply:
Acked